src/main/java/com/aphyr/riemann/client/Promise.java

```diff
@@ -75,6 +99,11 @@ public Object unsafeDeref(final long time,
     }
   }
 
+  @Override
+  public void setCallback(Callback<T> c) {
```

Comment:
Looks like a race condition: callbacks won't be triggered if their values are delivered early.